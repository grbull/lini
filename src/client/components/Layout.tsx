import cn from 'classnames';
import React, { ReactElement, ReactNode } from 'react';

import { Footer } from './Footer';
import { Navigation } from './Navigation';

interface Props {
  children: ReactNode;
}

export function Layout({ children }: Props): ReactElement {
  return (
    <>
      <main
        className={cn(
          'pb-14 md:pb-10',
          'pt-0 md:pt-18',
          'md:max-w-4xl',
          'mx-0 md:mx-auto'
        )}
      >
        {children}
      </main>
      <Navigation />
      <Footer />
    </>
  );
}
