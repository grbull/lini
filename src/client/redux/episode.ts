/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
