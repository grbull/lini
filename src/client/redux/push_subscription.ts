/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { PushSubscriptionCreateDto } from '../../server/push_subscription/push_subscription.dto';
import { api } from '../utils/api';

type PushSubscriptionState = {
  status: 'loading' | 'idle' | 'error';
  error?: string;
};

const initialState: PushSubscriptionState = {
  status: 'idle',
};

export const create = createAsyncThunk<void, PushSubscriptionCreateDto>(
  'push_subscription/get',
  (createDto) => api.push.create(createDto)
);

export const pushSubscription = createSlice({
  name: 'push_subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state) => {
      state.status = 'idle';
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const pushSubscriptionActions = {
  ...pushSubscription.actions,
  create,
};
