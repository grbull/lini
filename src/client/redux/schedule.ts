/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  EpisodeDto,
  EpisodeScheduleDto,
} from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type ScheduleState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: {
    future: EpisodeDto[];
    past: EpisodeDto[];
  };
  error?: string;
};

const initialState: ScheduleState = {
  status: 'init',
  data: {
    future: [],
    past: [],
  },
};

export const get = createAsyncThunk<EpisodeScheduleDto>('schedule/get', () =>
  api.episode.getSchedule()
);

export const schedule = createSlice({
  name: 'schedule',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const scheduleActions = {
  ...schedule.actions,
  get,
};
