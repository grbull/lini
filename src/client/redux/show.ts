/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowWithEpisodesDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type ShowState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data?: ShowWithEpisodesDto;
  error?: string;
};

const initialState: ShowState = {
  status: 'init',
};

export const get = createAsyncThunk<ShowWithEpisodesDto, number>(
  'show/get',
  (id) => api.show.get(id)
);

export const show = createSlice({
  name: 'show',
  initialState,
  reducers: {
    reset: () => {
      return initialState;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.status = 'loading';
      state.data = undefined;
      state.error = undefined;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const showActions = {
  ...show.actions,
  get,
};
