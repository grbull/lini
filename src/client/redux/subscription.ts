/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import {
  SubscriptionCreateDto,
  SubscriptionDto,
  SubscriptionRemoveDto,
} from '../../server/subscription/subscription.dto';
import { api } from '../utils/api';
import { scheduleActions } from './schedule';

type SubscriptionState = {
  status: 'init' | 'loading' | 'idle' | 'error';
  data: SubscriptionDto[];
  error?: string;
};

const initialState: SubscriptionState = {
  status: 'init',
  data: [],
};

export const getAll = createAsyncThunk<SubscriptionDto[]>(
  'subscription/getAll',
  () => api.subscription.getAll()
);

export const create = createAsyncThunk<SubscriptionDto, SubscriptionCreateDto>(
  'subscription/create',
  async (createDto, thunkAPI) => {
    const subscription = await api.subscription.create(createDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const remove = createAsyncThunk<SubscriptionDto, SubscriptionRemoveDto>(
  'subscription/remove',
  async (removeDto, thunkAPI) => {
    const subscription = await api.subscription.remove(removeDto);
    thunkAPI.dispatch(scheduleActions.get());
    return subscription;
  }
);

export const subscription = createSlice({
  name: 'subscription',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getAll.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(getAll.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(getAll.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(create.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(create.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.push(payload);
    });
    builder.addCase(create.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });

    builder.addCase(remove.pending, (state) => {
      state.status = 'loading';
      state.error = undefined;
    });
    builder.addCase(remove.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.splice(
        state.data.findIndex((s) => s.show.id === payload.show.id)
      );
    });
    builder.addCase(remove.rejected, (state, { error }) => {
      state.status = 'error';
      state.error = error.message;
    });
  },
});

export const subscriptionActions = {
  ...subscription.actions,
  getAll,
  create,
  remove,
};
