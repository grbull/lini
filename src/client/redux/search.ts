/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { ShowDto } from '../../server/show/show.dto';
import { api } from '../utils/api';

type SearchState = {
  isLoading: boolean;
  error?: string;
  data?: ShowDto[];
};

const initialState: SearchState = {
  isLoading: false,
};

export const query = createAsyncThunk<ShowDto[], string>(
  'search/query',
  (query) => api.show.search(query)
);

export const search = createSlice({
  name: 'search',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(query.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(query.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(query.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const searchActions = { ...search.actions, query };
