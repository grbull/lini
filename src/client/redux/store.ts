import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import { episode } from './episode';
import { pushSubscription } from './push_subscription';
import { schedule } from './schedule';
import { search } from './search';
import { show } from './show';
import { subscription } from './subscription';
import { user } from './user';

export const reducers = combineReducers({
  user: user.reducer,
  search: search.reducer,
  subscription: subscription.reducer,
  show: show.reducer,
  episode: episode.reducer,
  schedule: schedule.reducer,
  pushSubscription: pushSubscription.reducer,
});

const persistConfig = {
  key: 'root',
  storage,
  version: 2,
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
  reducer: persistedReducer,
  // https://github.com/rt2zz/redux-persist/issues/988#issuecomment-552242978
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  devTools: process.env.NODE_ENV === 'development',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
