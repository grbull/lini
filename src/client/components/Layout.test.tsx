/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';

import { reducers } from '../redux/store';
import { Layout } from './Layout';

describe('Layout Component', () => {
  it('matches the snapshot', () => {
    const history = createMemoryHistory();

    const store = configureStore({ reducer: reducers });

    window.scrollTo = jest.fn();

    const { asFragment } = render(
      <Router history={history}>
        <Provider store={store}>
          <Layout>Generic children</Layout>
        </Provider>
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
