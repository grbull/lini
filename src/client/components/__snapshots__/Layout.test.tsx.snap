// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component matches the snapshot 1`] = `
<DocumentFragment>
  <main
    class="pb-14 md:pb-10 pt-0 md:pt-18 md:max-w-4xl mx-0 md:mx-auto"
  >
    Generic children
  </main>
  <div />
  <nav
    class="h-14 fixed bottom-0 md:top-0 w-full bg-gray-200 dark:bg-gray-800"
  >
    <div
      class="h-14 w-full md:max-w-4xl mx-0 md:mx-auto md:flex md:justify-between md:items-center md:px-2.5"
    >
      <a
        class="hidden md:block font-semibold text-2xl"
        href="/"
      >
        Lini
      </a>
      <div
        class="h-full flex justify-around items-center"
      >
        <a
          class="w-full h-full flex flex-col md:flex-row justify-center items-center text-sm focus:outline-none md:ml-4 md:px-2 whitespace-nowrap bg-gray-300 text-gray-800 dark:bg-gray-800 dark:text-white"
          href="/"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-home fa-w-18 m-1 text-lg md:mr-1"
            data-icon="home"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
              fill="currentColor"
            />
          </svg>
          Home
        </a>
        <a
          class="w-full h-full flex flex-col md:flex-row justify-center items-center text-sm focus:outline-none md:ml-4 md:px-2 whitespace-nowrap bg-gray-200 text-gray-600 dark:bg-gray-800 dark:text-gray-400"
          href="/search"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-search fa-w-16 m-1 text-lg md:mr-1"
            data-icon="search"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              fill="currentColor"
            />
          </svg>
          Search
        </a>
        <a
          class="w-full h-full flex flex-col md:flex-row justify-center items-center text-sm focus:outline-none md:ml-4 md:px-2 whitespace-nowrap bg-gray-200 text-gray-600 dark:bg-gray-800 dark:text-gray-400"
          href="/library"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-folder-open fa-w-18 m-1 text-lg md:mr-1"
            data-icon="folder-open"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.694 292.093L500.27 416.248A63.997 63.997 0 0 1 444.989 448H45.025c-18.523 0-30.064-20.093-20.731-36.093l72.424-124.155A64 64 0 0 1 152 256h399.964c18.523 0 30.064 20.093 20.73 36.093zM152 224h328v-48c0-26.51-21.49-48-48-48H272l-64-64H48C21.49 64 0 85.49 0 112v278.046l69.077-118.418C86.214 242.25 117.989 224 152 224z"
              fill="currentColor"
            />
          </svg>
          Your Library
        </a>
      </div>
    </div>
  </nav>
  <div
    class="h-8 fixed bottom-0 hidden md:block w-full bg-gray-100 dark:bg-gray-600"
  >
    <div
      class="max-w-4xl h-full mx-auto flex items-center justify-center font-light text-sm"
    >
      <span>
        Lini 2021 ©
      </span>
      <span
        class="px-4"
      >
        •
      </span>
      <span>
        Version: 2.11.0
      </span>
      <span
        class="px-4"
      >
        •
      </span>
      <a
        class="underline"
        href="https://github.com/grbull/lini"
      >
        Source
      </a>
    </div>
  </div>
</DocumentFragment>
`;
